### DISCUSSION TITLE
Added: Notes for September 30th, 2020

### DISCUSSION THREAD

**Discussion by 333fred** [dotnet/csharplang#3959] [LDM Notes]
https://github.com/dotnet/csharplang/blob/master/meetings/2020/LDM-2020-09-30.md

## Agenda

- `record structs`
    - `struct` equality
    - `with` expressions
    - Primary constructors and `data` properties

**Comment by MgSam** [comment]
> Given that, we'd rather work with the runtime team to make the automatic Equals method better, which will benefit not just all C# structs, but all struct types from all languages that run on the CLR.

This sounds great, but is there a proposal for how to actually do this? If the C# compiler isn't generating Equals then what option do you have at runtime besides reflection?

> In a sense, a record struct would just become a strongly-named tuple type, and have all the same behaviors as a standard tuple type.

+1 for record structs having mutable public properties like tuples. Not only does it give you a tuple with a name, but it would also solve the problem of tuples not being reflection friendly because the property names are compile time only.

**Reply by CyrusNajmabadi** [reply]
> This sounds great, but is there a proposal for how to actually do this? If the C# compiler isn't generating Equals then what option do you have at runtime besides reflection?

Runtime knows the information about the type and generate an optimized version per struct on demand and then use that.  Similar to how Enum operatoins used to be slow and boxy, but now can be fast and integral oriented when the runtime specializes them.

**Reply by CyrusNajmabadi** [reply]
>  but it would also solve the problem of tuples not being reflection friendly because the property names are compile time only.

Great point.  Will bring that up in our next meeting on it :)

**Reply by ufcpp** [reply]
rather public fields like ValueTuple than like Tuple?

**Reply by 333fred** [reply]
> rather public fields like ValueTuple than like Tuple?

Yes, `ValueTuple`.

**Comment by YairHalberstadt** [comment]
I opened a proposal on dotnet runtime recently suggesting that tiered compilation be used to optimise struct equality if used on a hot path: https://github.com/dotnet/runtime/issues/36613

**Comment by orthoxerox** [comment]
Will `with` work on `readonly struct`s?

Also, -1 for making positional record structs mutable. Different behavior between record classes and structs will be too confusing, especially if you add primary constructors to regular structs and classes later.

**Reply by 333fred** [reply]
> Will `with` work on `readonly struct`s?

I don't see why it wouldn't. Init properties are allowed there, so it should be fine.
